qdb_dask_connector/__init__.py

```diff
@@ -0,0 +1,286 @@
+    match = re.search(range_pattern, query)
+    # first we check try to extract "in range (start, end)" from query
+    # if we can't do it we will query first() and last() from the table
+    query_range = tuple()
+    if match:
+        start_str = match.group(1)
+        end_str = match.group(2)
+        logger.debug("Extracted strings: (%s, %s)", start_str, end_str)
+        parser_settings = {
+            "PREFER_DAY_OF_MONTH": "first",
+            "PREFER_MONTH_OF_YEAR": "first",
+        }
+        start_date = dateparser.parse(start_str, settings=parser_settings)
+        end_date = dateparser.parse(end_str, settings=parser_settings)
+        query_range = (start_date, end_date)
+        logger.debug("Parsed datetime: %s", query_range)
+    else:
+        logger.debug(
+            "No range found in query, querying table for first and last timestamp"
+        )
+        range_query = f"SELECT first($timestamp), last($timestamp) FROM {table_name}"
+        df = qdbpd.query(conn, range_query)
+        if not df.empty:
+            df.loc[0, "last($timestamp)"] += datetime.timedelta(microseconds=1)
+            query_range += tuple(df.iloc[0])
+            logger.debug("Extracted range from table: %s", query_range)
+
+    return query_range
+
+
+def _create_subrange_query(
+    query: str, query_range: tuple[datetime.datetime, datetime.datetime]
+) -> str:
+    # this function is used to mock part of future c api functionality
+    """
+    Adds range to base query.
+    IF range is found in the query, it will be replaced with the new range.
+    IF no range is found, it will be added after the "FROM {table}" clause.
+    """
+    new_query = query
+    range_match = re.search(range_pattern, query)
+    start_str = query_range[0].strftime("%Y-%m-%dT%H:%M:%S.%fZ")
+    end_str = query_range[1].strftime("%Y-%m-%dT%H:%M:%S.%fZ")
+    if range_match:
+        if len(range_match.groups()) == 2:
+            new_query = re.sub(
+                range_pattern,
+                f"IN RANGE ({start_str}, {end_str})",
+                query,
+            )
+        logger.debug("Created subquery: %s", new_query)
+        return new_query
+
+    table_match = re.search(table_pattern, query)
+    new_query = re.sub(
+        table_pattern,
+        f"FROM {table_match.group(1)} IN RANGE ({start_str}, {end_str})",
+        query,
+    )
+    logger.debug("Created subquery: %s", new_query)
+    return new_query
+
+
+def _get_subqueries(conn, query: str, table_name: str) -> list[str]:
+    # this function is used to mock part of future c api functionality
+    shard_size = conn.table(table_name.replace('"', "")).get_shard_size()
+    start, end = _extract_range_from_query(conn, query, table_name)
+    ranges_to_query = conn.split_query_range(start, end, shard_size)
+
+    subqueries = []
+    for rng in ranges_to_query:
+        subqueries.append(_create_subrange_query(query, rng))
+    return subqueries
```

Comment:
Don't call this "subqueries", that is inherently going to cause infinite confusion as subqueries in SQL are an entirely different thing.

I recommend calling this function "_split_query".